cpp/src/copying/contiguous_split.cu

```diff
@@ -194,11 +194,16 @@ __device__ void copy_buffer(uint8_t* __restrict__ dst,
     if (value_shift || bit_shift) {
       std::size_t idx = (num_bytes - remainder) / 4;
       uint32_t v = remainder > 0 ? (reinterpret_cast<uint32_t const*>(src)[idx] - value_shift) : 0;
+      auto const have_trailing_bits = ((num_elements * 32) - num_rows) < bit_shift;
       while (remainder) {
-        uint32_t const next = bit_shift > 0 || remainder > 4
-                                ? (reinterpret_cast<uint32_t const*>(src)[idx + 1] - value_shift)
-                                : 0;
-        uint32_t const val  = (v >> bit_shift) | (next << (32 - bit_shift));
+        // if we're at the very last word of a validity copy, we do not always need to read the next
+        // word to get the final trailing bits.
+        auto const read_trailing_bits = bit_shift > 0 && remainder == 4 && have_trailing_bits;
```

Comment:
Is the 4 here `sizeof(uint32_t)`? Not requesting a change, just trying to understand the logic.

Reply:
Correct.  We're reading (up to) the trailing 15 bytes of the buffer to be copied.  In the case where the buffer happens to be validity, the elements are all `bitmask_type` words.

The fundamental issue is that if we're copying from some arbitrary row, we have to shift the bits of any validity around.  So imagine we're reading 1 bit starting at row 31.  That's all within the "final" word - we just need to shift that bit up by 31 for the output.   But let's say we want to read 2 bits starting at row 31.  The first bit comes from the word we just read, but the 2nd bit comes from the next word (idx + 1) - so we have to read it.   

The initial bug here was that we were doing this when we shouldn't have been.

Reply:
Apologies in advance for the drive-by review comment. Is there any reason not to define this variable? I see multiple magic numbers throughout this code (lots of 4s and 32s) and it seems like a `constexpr auto uint32_size = sizeof(uint32_t);` would help avoid questions like this.

Reply:
> drive-by review comment

:joy: 

I think it would be good, if not a bit out of scope of the PR. Up to the author IMO.

Reply:
I think something like this definitely adds clarity.

```
constexpr size_type rows_per_element = 32;
auto const have_trailing_bits = ((num_elements * rows_per_element) - num_rows) < bit_shift;
```

But I also think this makes more sense as is, since this reads as pretty standard bit-shifting stuff. 
`uint32_t const val = (v >> bit_shift) | (next << (32 - bit_shift));`

Changing it to this would obfuscate I think.  
`uint32_t const val = (v >> bit_shift) | (next << (rows_per_bitmask - bit_shift));`